environment_switcher/lib/src/environment_switcher.dart

```diff
@@ -0,0 +1,195 @@
+  final bool showBanner;
+
+  /// A banner that visually shows the user what [Environment]
+  /// is currently set to.
+  /// This is useful because it's possible to be on one Flavour
+  /// in the native code, and another in the Flutter code.
+  const EnvironmentSwitcher({
+    @required this.builder,
+    @required this.environments,
+    this.environmentStore, // mainly required for testing
+    this.showBanner = true,
+    this.defaultEnvironment,
+  })  : assert(builder != null),
+        assert(environments != null && environments.length != 0),
+        assert(showBanner != null),
+        assert(showBanner ? defaultEnvironment != null : true);
+
+  @override
+  State<StatefulWidget> createState() {
+    return _StateEnvironmentSwitcher(environments, environmentStore);
+  }
+}
+
+class _StateEnvironmentSwitcher extends State<EnvironmentSwitcher> {
+  final List<Environment> environments;
+  final EnvironmentStore environmentStore;
+
+  Environment currentEnvironment;
+
+  Environment get firstEnvironmentOrDefault {
+    if ((environments ?? []).isNotEmpty) {
+      return environments[0];
+    } else {
+      return widget.defaultEnvironment;
+    }
+  }
+
+  _StateEnvironmentSwitcher(this.environments, EnvironmentStore store)
+      : environmentStore = store ?? EnvironmentStore(store: Store());
+
+  @override
+  Widget build(BuildContext context) {
+    // If we don't want to show the banner, we can hide it altogether easily.
+    // When this happens, we want to show the default environment no matter the
+    // previously saved environment
+    // Note: Adding "== false" also acts as a null-check.
+    if (widget.showBanner == false) {
+      return widget.builder?.call(widget.defaultEnvironment);
+    }
+
+    if (currentEnvironment == null) {
+      // On first run, we need to load the saved environment (if any). If none
+      // are saved, use the first in the list. If the list is empty, use the
+      // default.
+      return FutureBuilder<Environment>(
+        future: _getSavedEnvironmentOrDefault(),
+        builder: (context, snapshot) {
+          currentEnvironment = snapshot.data ?? firstEnvironmentOrDefault;
+          return _Banner(
+            environment: currentEnvironment,
+            child: widget.builder?.call(currentEnvironment),
+            onBannerTapped: _onBannerTapped,
+          );
+        },
+      );
+    } else {
+      return _Banner(
+        environment: currentEnvironment,
+        child: widget.builder?.call(currentEnvironment),
+        onBannerTapped: _onBannerTapped,
+      );
+    }
+  }
+
+  Future<Environment> _getSavedEnvironmentOrDefault() async {
+    try {
+      final savedEnvironmentName = await environmentStore.getSavedEnvironment();
+      if (savedEnvironmentName == null) {
+        throw "No environment saved";
+      }
+      final environment =
+          environments.firstWhere((env) => env.name == savedEnvironmentName, orElse: () => null);
```

Comment:
Is matching on name correct? Are we doing any asserts to check that name is always unique? Is value more appropriate?

Reply:
I agree this is a bit questionable. I reckoned it's highly unlikely (also bad practice) that two environments would have the same name. Short of creating a way of converting the object to string and back (or use `JsonSerialization`), this would have to do.